library/core/src/intrinsics.rs

```diff
@@ -1303,6 +1303,16 @@ extern "rust-intrinsic" {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     pub fn arith_offset<T>(dst: *const T, offset: isize) -> *const T;
 
+    #[cfg(not(bootstrap))]
+    /// The offset of the `Some(_)` value of an `Option<T>`. Used internally for
+    /// `Option::as_slice` and `Option::as_mut_slice`. This needs to be called with an
+    /// `Option<_>` type.
+    // FIXME: This should be replaced once we get a stable public method for getting
+    // the offset of enum variant's fields (e.g. an extension of `offset_of!` to enums)
+    #[rustc_const_stable(feature = "const_option_some_offset", since = "1.68.0")]
+    #[rustc_safe_intrinsic]
+    pub fn option_some_offset<T>() -> usize;
```

Comment:
Pondering: using the raw offset made sense when we were doing the hack, but it's rather indirect if we're adding an intrinsic to do what we want.

What if, instead, it was just
```rust
pub fn option_payload_ptr<T>(p: *const Option<T>) -> *const T;
```
which could be put in `LowerIntrinsics` to turn into
```rust
&raw (((*p) as Some).0: T)
```
in MIR, and then the backends wouldn't need to know about it.

(AKA no change to `rustc_codegen_ssa`, nor to `rustc_codegen_cranelift` or `rustc_codegen_gcc`, which I think are broken by this PR as currently written, if anyone uses `Option::as_slice` on them.)

Then the implementation for `as_slice` would be straight-forward, just `from_raw_parts(option_payload_ptr(self), usize::from(self.is_some()))`.  And it would be sound for `as_mut_slice` to use that too, with some pointer casting.

And it avoids ever needing to look at layout information directly, since all we really wanted was a pointer projection.  (And someone who wants the offset can do a subtraction themselves to get it, like the offset_of crate does for *all* offsets.)

Reply:
Ok, I'm a bit wary about the pointer casting stuff, but if you say so, that will certainly be easier.

Reply:
Well, MIRI should catch it if you accidentally get the stacked borrows wrong, but it shouldn't be a big deal to make it work, since casting between `*const` and `*mut` is always fine.

Thus AFAIK this will work:
```rust
fn as_mut_slice(&mut self) -> &mut [T] {
    let len = usize::from(self.is_some());
    let ptr: *mut Self = self; // Make sure to use the mutable borrow's lease, not `&*self as *const _`.
    unsafe {
        let payload_ptr = option_payload_ptr(ptr.cast_const()).cast_mut();
        from_raw_parts_mut(payload_ptr, len)
    }
}
```
where I've intentionally _never_ used `as`, in order to make sure I'm always getting _exactly_ the conversion I'm intending.

And the `*const` is only used for field projection, where `*mut` vs `*const` definitely doesn't matter, since you get a pointer back with the same provenance, and we never introduced any intermediate borrows that could affect the stacked borrows leases.

(I probably have some of the terms here wrong, but I'm pretty confident in the overall thrust of things.  If Ralf contradicts anything I said here, listen to him instead 😅)

Reply:
This is just to avoid having two intrinsic for const and mut? Sure makes sense.